util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+int ft_client_connect(struct ct_pingpong *ct);
+int ft_alloc_active_res(struct ct_pingpong *ct, struct fi_info *fi);
+int ft_init_ep(struct ct_pingpong *ct);
+int ft_av_insert(struct fid_av *av, void *addr, size_t count, fi_addr_t *fi_addr,
+		uint64_t flags, void *context);
+int ft_init_av(struct ct_pingpong *ct);
+void ft_free_res(struct ct_pingpong *ct);
+void init_test(struct ct_pingpong *ct, struct ft_opts *opts, char *test_name, size_t test_name_len);
+
+//int ft_sync(struct ct_pingpong *ct);
+int ft_finalize(struct ct_pingpong *ct);
+
+ssize_t ft_post_rx(struct ct_pingpong *ct, struct fid_ep *ep, size_t size, struct fi_context* ctx);
+ssize_t ft_post_tx(struct ct_pingpong *ct, struct fid_ep *ep, size_t size, struct fi_context* ctx);
+ssize_t ft_rx(struct ct_pingpong *ct, struct fid_ep *ep, size_t size);
+ssize_t ft_tx(struct ct_pingpong *ct, struct fid_ep *ep, size_t size);
+ssize_t ft_inject(struct ct_pingpong *ct, struct fid_ep *ep, size_t size);
+
+int ft_cq_readerr(struct fid_cq *cq);
+int ft_get_rx_comp(struct ct_pingpong *ct, uint64_t total);
+int ft_get_tx_comp(struct ct_pingpong *ct, uint64_t total);
+
+void eq_readerr(struct fid_eq *eq, const char *eq_str);
+
+int64_t get_elapsed(const struct timespec *b, const struct timespec *a,
+		enum precision p);
+void show_perf(char *name, int tsize, int sent, int acked, struct timespec *start,
+		struct timespec *end, int xfers_per_iter);
+
+int ft_getsrcaddr(char *node, char *service, struct fi_info *hints);
+int ft_read_addr_opts(struct ct_pingpong *ct, char **node, char **service, struct fi_info *hints,
+		uint64_t *flags, struct ft_opts *opts);
+char *size_str(char str[FT_STR_LEN], long long size);
+char *cnt_str(char str[FT_STR_LEN], long long cnt);
+int size_to_count(struct ct_pingpong *ct, int size);
+
+int pingpong(struct ct_pingpong *ct);
+
+void ft_init_ct_pingpong(struct ct_pingpong *ct);
+
+/*******************************************************************************************/
+/*                                       FT func                                           */
+/*******************************************************************************************/
+
+int ft_ctrl_init(struct ct_pingpong  *ct)
+{
+	int err, ret;
+
+	if (ct->opts.dst_addr) {
+		ct->ctrl_connfd = socket(AF_INET, SOCK_STREAM, 0);
+		if (ct->ctrl_connfd == -1) {
+			err = errno;
+			FT_PRINTERR("socket", err);
+			return ct->ctrl_connfd;
+		}
+
+		memset(&ct->ctrl_addr, '\0', sizeof(ct->ctrl_addr));
+		ct->ctrl_addr.sin_family = AF_INET;
+		ret = inet_pton(AF_INET, ct->opts.dst_addr, &(ct->ctrl_addr.sin_addr));
+		if (ret == 0) {
+			err = errno;
+			FT_PRINTERR("inet_pton", err);
+			return ret;
+		}
+		ct->ctrl_addr.sin_port = htons(ct->ctrl_port);
+
+		FT_DEBUG("CLIENT: connecting to <%s>\n", ct->opts.dst_addr);
+		ret = connect(ct->ctrl_connfd, (struct sockaddr *)&ct->ctrl_addr, sizeof(ct->ctrl_addr));
+		if (ret == -1) {
+			err = errno;
+			FT_PRINTERR("connect", err);
```

Comment:
The `FT_PRINTERR` function expects a negative return value, but this errno value is positive. This results in something like: 

```
connect(): util/pingpong.c:381, ret=111 (Unknown error 18446744073709551505)
```

So make sure that every instance where `FT_PRINTERR` is called, it is called with a negative number. Or modify `FT_PRINTERR` to accept positive values and negate them at the call site.